### PR TITLE
refactor(core): move `validate_memory_params` to `AllocationsCotntext…

### LINKED ISSUE #3813: move `validate_memory_params` to `AllocationsCotntext::new`
### File Location(s)

core-processor/src/executor.rs
core/src/memory.rs

### Proposal

Presently we do not check `AllocationsContext` for values consistency in `new(...)`, but this should be done. See  `validate_memory_params ` to understand what checks should be added

Also `IncorrectAllocationDataError` can be removed after that, because AllocationsContext will guaranties that this is impossible.

### PR DESCRIPTION
**Pull request by ByteNacked**
Resolves #3813 .

Move `validate_memory_params` into `AllocationsContext` constructor and make it capable of failing.

Some tests in core/src/memory.rs were updated due to new invariants established by the validation of memory parameters upon `AllocationsContext` creation.

@reviewer-or-team

### CODE REVIEW

File: core/src/memory.rs
Comment (ByteNacked) [inline core/src/memory.rs:442]: This invariant should be upheld by `valdidate_memory_params` upon creation of object.

File: core/src/pages.rs
Comment (grishasobol) [reply core/src/pages.rs:59]: It's not safe to do this, because `gear-core` is big enough, so developers can simply make bugs using this constructor. `PagesAmount` supports `Into<u32>`, `From<u16>` and `TryFrom<u32>` this must be enough for any purposes in tests.

File: core/src/pages.rs
Comment (grishasobol) [inline core/src/pages.rs:189]: The same  unsafe

File: core/src/pages.rs
Comment (grishasobol) [reply core/src/pages.rs:59]: Or you can make it `pub` for tests only

**Review by grishasobol — CHANGES_REQUESTED** [review]

File: core/src/pages.rs
Comment (ByteNacked) [reply core/src/pages.rs]: Fixed, added method `[cfg(test)] fn into_inner() -> u32`

File: core/src/memory.rs
Comment (ByteNacked) [reply core/src/memory.rs]: The problem with the slow property test was the poor performance of the  `AllocationContext::alloc` with opt = 1 in dev profile in happy case.

```rust
pub fn alloc<G: GrowHandler>(
{
      ...
        self.allocations.extend(interval.iter()); << HERE
       ...
}
```

For some reason, `IntervalIterator::next` works very slowly with opt = 1 in the development profile. I'm not sure if it should be fixed, but I made a workaround by reducing the chance of successful allocation in the prop-test (Same behavior as it was before this PR).

File: core/src/pages.rs
Comment (grishasobol) [reply core/src/pages.rs]: I cannot understand what the reason to add this method ? you can use `u32::from(page)`

File: core/src/memory.rs
Comment (ByteNacked) [reply core/src/memory.rs]: In the previous version of the property test, `AllocationContext::alloc` rarely succeeds due to the way its strategy generates input parameters. (Almost never executes `self.allocations.extend(interval.iter());`) 

After I changed the strategy, `AllocationContext::alloc`  started to succeeded more frequently (thus executing `self.allocations.extend(interval.iter());` more frequently) . That led to prop-test slow down.

Now it fixed by prop-test's strategy generating input parameters in such way which decrease chance of `AllocationContext::alloc` succession. (Same behavior as before PR).

File: core/src/pages.rs
Comment (ByteNacked) [reply core/src/pages.rs]: Missed it, fixed

File: core/src/memory.rs
Comment (ByteNacked) [reply core/src/memory.rs]: it seems that actual problem was in `BTree::extend` call.
Confirmed this with a snippet:
```rust
        //self.allocations.extend(interval.iter());
        for i in interval.iter() {
            core::hint::black_box(i);
        }
```

After discussion changed in prop-test number of generated pages in action per allocation. 
```rust
        fn actions() -> impl Strategy<Value = Vec<Action>> {
            let action = prop_oneof![
                // Allocate smaller number (0..32) of pages due to `BTree::extend` significantly slows down prop-test.
                wasm_pages_amount_with_range(0, 32).prop_map(|pages| Action::Alloc { pages }),
        ....
```
It solves problem with prop-test slow down too.

File: core/src/memory.rs
Comment (grishasobol) [inline core/src/memory.rs:846]: Please append new struct here `MemoryParams`

**Review by grishasobol — APPROVED** [review]
Ok. Everything looks good. But before commit I would ask you to make local run of memory property tests with bigger amount of iterations in order to check everything is fine there

File: core/src/memory.rs
Comment (ByteNacked) [reply core/src/memory.rs:846]: I have a similar idea, but it seems impossible to create a combined strategy that uses a `struct` for `Value` instead of a tuple. The `prop_flat_map` combinator transforms a tuple of strategies into a strategy of tuples, but it cannot do the same for a struct.

Comment (grishasobol) [reply core/src/memory.rs:846]: Hm, yes. May be the most correct way is to implement `Arbitrary` for `struct MemoryParams`.
You can see examples in common/numerated/src/mock.rs. But may be better to make it in other PR

Comment (ByteNacked) [reply core/src/memory.rs:846]: Created issue #3906 3906

**Comment by ByteNacked** [comment]
Waiting for 1 more reviewer's approval

**Comment by grishasobol** [comment]
So, I think this review hangs on enough of time and can be merged

**Pull request merged**
